### PR TITLE
fix(configure): preserve custom config keys across wizard sections

### LINKED ISSUE #9632: openclaw configure wizard overwrites entire config, dropping custom keys
## Bug

Running `openclaw configure` rewrites the entire `openclaw.json` file, dropping any custom settings that aren't managed by the wizard.

## Steps to Reproduce

1. Set custom config values via `openclaw config set`, e.g.:
   - `tools.media.audio` (pinned transcription model)
   - `gateway.auth.allowTailscale`
   - Disabled skills entries (e.g. 22 disabled skills)
2. Run `openclaw configure` (e.g. to add a new model auth)
3. Observe that the written config no longer contains the custom keys from step 1

## Expected Behavior

`openclaw configure` should preserve any config keys it doesn't explicitly manage. The wizard should merge its changes into the existing config rather than replacing the entire file.

## Actual Behavior

The wizard reads the config into `baseConfig`, spreads it into `nextConfig`, but only the sections the user interacts with get carried forward. Keys that the wizard doesn't have UI for (like `tools.media.audio`, skill disable states, custom gateway auth flags) are silently dropped when `writeConfigFile(nextConfig)` is called.

## Workaround

Use `openclaw config set` / `openclaw config unset` for granular changes, and `openclaw models auth login --provider <name>` for auth changes. Avoid the full `openclaw configure` wizard after initial setup.

## Environment

- OpenClaw version: 2026.2.2-3
- OS: Ubuntu 24.04 (VPS)
- Node: v24.13.0

### PR DESCRIPTION
**Pull request by mcaxtr**
## Summary

Fixes #9632

The configure wizard was overwriting the entire config file when modifying a single section, dropping custom keys the wizard doesn't manage (e.g., `tools.media.audio`, `gateway.auth.allowTailscale`, disabled `skills.entries`).

**Root cause**: The wizard reads config through `readConfigFileSnapshot()`, which returns a Zod-validated + defaults-applied version. This processed config was used as the base for writes, meaning:
- Zod-processed sections replaced raw file values (adding defaults the user never set)
- `${VAR}` env references were substituted with literal values
- `$include` directives were resolved to inline content

**Fix**: Added `mergeWizardOutput()` which uses reference equality between `baseConfig` and `nextConfig` to detect which top-level sections the wizard actually modified. Only wizard-modified sections use the wizard's output; unmodified sections preserve their raw file values from `snapshot.parsed`.

- Unmodified sections keep raw file values (no injected defaults)
- `${VAR}` references preserved in unmodified sections
- `$include` at top level falls back to current behavior (separate concern)
- Both local and remote wizard paths use the merge function

## Test plan

- [x] `mergeWizardOutput` unit tests: preserves raw values, uses wizard values for modified keys, preserves keys not in nextConfig, falls back with `$include`, adds new wizard keys
- [x] Integration tests: preserves tools/skills/logging when configuring only `model`, `gateway`, or interactive loop
- [x] Edge case: handler that drops keys → recovered from rawParsed
- [x] Edge case: rawParsed differs from baseConfig (minimal raw vs defaults-injected) → unmodified sections use raw values
- All 11 tests fail before fix, pass after

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR updates the configure wizard to avoid rewriting the entire config when only one wizard section changes. It introduces `mergeWizardOutput()` and uses it in both local/remote wizard flows so that top‑level sections the wizard didn’t modify are preserved from the raw parsed file (keeping user custom keys, `${VAR}` references, and un-expanded `$include` content), while modified sections come from the wizard output. The accompanying tests expand coverage for preserving raw values, recovering keys dropped by section handlers, and `$include` fallback behavior.

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with minimal risk.
- Changes are narrowly scoped to how the wizard merges its output back into the raw parsed config, with explicit handling for default-only sections and `$include` fallback; unit/integration tests cover the key regressions described in the PR.
- No files require special attention

### CODE REVIEW

**Comment by mcaxtr** [comment]
@greptile please re-review this PR.

File: src/commands/configure.wizard.ts [hunk 1]
```diff
@@ -43,6 +43,37 @@ import {
...
+
+  const result: Record<string, unknown> = { ...rawParsed };
+  const baseRec = baseConfig as unknown as Record<string, unknown>;
+  const nextRec = nextConfig as unknown as Record<string, unknown>;
+
+  for (const key of Object.keys(nextRec)) {
+    if (nextRec[key] !== baseRec[key]) {
+      result[key] = nextRec[key];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/commands/configure.wizard.ts:81]: This behavior is intentional and correct. Here's the full trace:

1. **All top-level schema keys are optional**: `AgentsSchema` at `zod-schema.agents.ts:12` ends with `.optional()`. Same for `tools`, `skills`, `logging`, etc. A config file with these keys absent passes validation.

2. **Zod re-injects defaults on next read**: When `readConfigFileSnapshot()` parses the file, Zod applies default values for missing optional sections. So default-only sections are never "permanently deleted" — they're ephemeral defaults that exist only in the in-memory config object.

3. **The user never set these keys**: If `agents` is in `baseConfig` but not in `rawParsed`, it means the user's config file never contained `agents` — Zod injected it during read. Writing it back would be the original bug (#9632): bloating the file with defaults the user never chose.

**Added in this push**:
- Explicit `expect(written.agents).toBeUndefined()` assertion in the test at line 397, confirming the behavior is tested and intentional
- JSDoc comment on `mergeWizardOutput` documenting why default-only sections are omitted

**Comment by mcaxtr** [comment]
@greptile please re-review this PR.

File: src/commands/configure.wizard.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/commands/configure.wizard.ts:81]: Fixed in 230eb47. The `mergeWizardOutput` function now has an `else if (!(key in rawParsed))` branch that carries forward default-only sections from `baseConfig`:

```typescript
} else if (!(key in rawParsed)) {
  // Section exists only in baseConfig (e.g., Zod defaults) and the
  // wizard didn't touch it — carry forward to avoid dropping it.
  result[key] = baseRec[key];
}
```

This ensures that sections injected by Zod defaults (like `agents`) are preserved even when they aren't in the raw file. Test updated to verify: `expect(written.agents).toEqual(baseConfig.agents)`.

**Comment by mcaxtr** [comment]
@greptile please re-review this PR.

File: src/commands/configure.wizard.ts [hunk 2]
```diff
@@ -43,6 +43,46 @@ import {
...
+ */
+export function mergeWizardOutput(
+  rawParsed: Record<string, unknown>,
+  baseConfig: OpenClawConfig,
+  nextConfig: OpenClawConfig,
+): Record<string, unknown> {
+  if ("$include" in rawParsed) {
+    return nextConfig as unknown as Record<string, unknown>;
+  }
+
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/commands/configure.wizard.ts:66]: Fixed in 3eceb39. Changed from `"\" in rawParsed` to `Object.hasOwn(rawParsed, "\")` to avoid prototype chain lookups. Good catch.

**Comment by mcaxtr** [comment]
@greptile please re-review this PR.

**Comment by steipete** [comment]
Closing as AI-assisted stale-fix triage.

Linked issue #9632 ("openclaw configure wizard overwrites entire config, dropping custom keys") is currently CLOSED and was closed on 2026-02-24T04:19:06Z with state reason NOT_PLANNED.
Given that issue state, this fix PR is no longer needed in the active queue and is being closed as stale.

If the underlying bug is still reproducible on current main, please reopen this PR (or open a new focused fix PR) and reference both #9632 and #13849 for fast re-triage.
